### ISSUE TITLE
Is there a recommended fix for error Type with the DIET Classifier?

### ISSUE THREAD

**Issue by jeff-ridgeway** [RasaHQ/rasa#8609]
**Rasa version**: 2.5.1

**Rasa SDK version** (if used & relevant):

**Rasa X version** (if used & relevant):

**Python version**: 3.7

**Operating system** (windows, osx, ...): Linux

**Issue**:

Right now, I am getting an error when I try launching rasa open-source with docker-compose up. The error(s) that I get include the following: `/opt/venv/lib/python3.7/site-packages/rasa/shared/utils/io.py:93: UserWarning: The model at 'model-location-here' could not be loaded. Error: <class 'ValueError'>: Wrong loss type 'cross_entropy', should be 'margin' or 'softmax'` and `/opt/venv/lib/python3.7/site-packages/rasa/shared/utils/io.py:93: UserWarning: Agent could not be loaded with the provided configuration. Load default agent without any model.` Has this error(s) been encountered before?

If so, what was the solution? I have a feeling that this is due to the DIET Classifier potentially, but I could be wrong. When testing this model outside of the instance that houses docker, the most recent model works no problem. I don't know if this could have an effect on anything (and it may be poor usage of git on my part) but I push a git repo (from AWS Sagemaker) with all of the rasa files that comes with the command `rasa init` as well as the models. I then pull the git repo into the EC2 instance and then try to run the models. I tried this back in December 2020 and had no issues.

**Error (including full traceback)**:

on EC2 instance where it is breaking: 
```
main2_rasa  | 2021-05-05 11:30:25 ERROR    rasa.core.agent  - Could not load model due to Wrong loss type 'cross_entropy', should be 'margin' or 'softmax'.
main2_rasa  | /opt/venv/lib/python3.7/site-packages/rasa/shared/utils/io.py:93: UserWarning: The model at 'beta2/models' could not be loaded. Error: <class 'ValueError'>: Wrong loss type 'cross_entropy', should be 'margin' or 'softmax'
main2_rasa  | /opt/venv/lib/python3.7/site-packages/rasa/shared/utils/io.py:93: UserWarning: Agent could not be loaded with the provided configuration. Load default agent without any model.
```

on AWS Sagemaker Instance where everything is fine and I run `rasa train`:
```
2021-05-05 11:24:50 INFO     rasa.model  - Data (version) for Core model section changed.
2021-05-05 11:24:50 INFO     rasa.model  - Data (version) for NLU model section changed.
2021-05-05 11:24:50 INFO     rasa.model  - Data (version) for Core model section changed.
2021-05-05 11:24:50 INFO     rasa.model  - Data (version) for NLU model section changed.
Training NLU model...
2021-05-05 11:24:51 INFO     rasa.nlu.utils.spacy_utils  - Trying to load spacy model with name 'en_core_web_md'
2021-05-05 11:24:53 INFO     rasa.nlu.components  - Added 'SpacyNLP' to component cache. Key 'SpacyNLP-en_core_web_md'.
/home/ec2-user/anaconda3/envs/JupyterSystemEnv/lib/python3.6/site-packages/rasa/utils/train_utils.py:565: UserWarning: constrain_similarities is set to `False`. It is recommended to set it to `True` when using cross-entropy loss. It will be set to `True` by default, Rasa Open Source 3.0.0 onwards.
  category=UserWarning,
/home/ec2-user/anaconda3/envs/JupyterSystemEnv/lib/python3.6/site-packages/rasa/utils/train_utils.py:537: UserWarning: model_confidence is set to `softmax`. It is recommended to try using `model_confidence=linear_norm` to make it easier to tune fallback thresholds.
  category=UserWarning,
```

**Command or request that led to error**:
```
sudo docker-compose up
```

**Content of configuration file (config.yml)** (if relevant):
```yml
not relevant as it works on AWS Sagemaker
```

**Content of domain file (domain.yml)** (if relevant):
```yml
not relevant as it works on AWS Sagemaker
```

**Comment by sara-tagger** [comment]
Thanks for the issue, @hsm207 will get back to you about it soon!

###### You may find help in the [docs](https://rasa.com/docs/) and the [forum,](https://forum.rasa.com/) too 🤗

**Comment by hsm207** [comment]
hi @jeff-ridgeway 

Are you sure the rasa version you used in AWS SageMaker, the  EC2 instance and in Docker compose is identical?

To check for the rasa version that was used to train the model, you can untar the model .tar.gz file and look at the metadata.json 
 file in the core and nlu folders for the rasa and rasa_version tag respectively, 

For example:

```bash
tar -xf 20210505-130354.tar.gz
echo `cat core/metadata.json | grep \"rasa\"`
echo `cat nlu/metadata.json | grep \"rasa_version\"`
```

If the model was trained with rasa 2.5.1, the the result of the two echo calls should be:

```bash
"rasa": "2.5.1"
"rasa_version": "2.5.1"
```

**Comment by jeff-ridgeway** [comment]
Hey @hsm207 

Thanks for getting back to me about this. I was able to run your commands on both my Sagemaker and AWS EC2 instance as well as my Dockerfile for `docker-compose`. Everything says 2.5.1 for the version. In my Dockerfile I have `rasa/rasa:2.5.1-full`. Seeing that have the right version, supposedly, do you have any of  places/things I should check for? 🤔

**Comment by hsm207** [comment]
hey @jeff-ridgeway 

I am not able to reproduce the error 🤔.  A couple of questions:

1. Is your environment back in Dec 2020 identical to your current environment or was there any update to AWS SageMaker, EC2 instance and/or the docker-compose?
2. Did you make any changes to `config.yml` after you ran `rasa init`?

**Comment by jeff-ridgeway** [comment]
1. I don't know if there was any update to AWS Sagemaker between last December and now. I didn't updated anything on my end. I just know that everything works on Sagemaker but not on the EC2 instance 🤷🏾‍♂️. The docker-compose file seems to be fine. This is my docker-compose file

```yml
version: '3.0'

services:
  rasa:
    container_name: rasa
    # go to docker hub/rasa changelog to see what version and flavour of rasa that you want
    # Make sure that the version you specify is the same as the version that you pip installed
    #image: rasa/rasa:2.5.1-full
    #image: flask-rasa:latest
    build:
      context: .
      dockerfile: Dockerfile
    expose:
      - port-number
    # This command will copy everything in current diretory to the /app directory in the container
    volumes:
      - .:/app
  duckling:
    image: rasa/duckling:latest
    ports:
      - "port:port"
      
  app:
    image: actions:latest
    ports:
      - "port-number1:port-number2"

  nginx:
    container_name: nginx
    image: nginx
    depends_on:
      - "rasa"
      - "duckling"
      - "app"
    ports:
      - port:port
      - port:port
    volumes:
      - path/to/directory
      - path/to/directory

```

2. Yes, I did make changes to the `config.yml` file after I ran `rasa init`. This is my current `config.yml` file:

```yml
# Configuration for Rasa NLU.
# https://rasa.com/docs/rasa/nlu/components/
language: en

pipeline:
# # No configuration for the NLU pipeline was provided. The following default pipeline was used to train your model.
# # If you'd like to customize it, uncomment and adjust the pipeline.
# # See https://rasa.com/docs/rasa/tuning-your-model for more information.
#    - name: WhitespaceTokenizer
#    - name: RegexFeaturizer
#    - name: LexicalSyntacticFeaturizer
   - name: "SpacyNLP"
     model: "en_core_web_md"
   - name: "SpacyTokenizer"
   - name: "SpacyFeaturizer"
     "pooling": "mean"
   - name: "SpacyEntityExtractor"
     dimensions: ["PERSON", "email"]
   - name: RegexEntityExtractor
     case_sensitive: False
     use_lookup_tables: True
   - name: DucklingEntityExtractor
     url: "http://duckling:port-number"
     dimensions: ["amount-of-money"]
     timeout : 3
     locale: "en_US"
   - name: DIETClassifier
     epochs: 150

# Configuration for Rasa Core.
# https://rasa.com/docs/rasa/core/policies/
policies:
# # No configuration for policies was provided. The following default policies were used to train your model.
# # If you'd like to customize them, uncomment and adjust the policies.
# # See https://rasa.com/docs/rasa/policies for more information.
  - name: MemoizationPolicy
  - name: TEDPolicy
    max_history: 5
    epochs: 100
  - name: RulePolicy
#   - form: FormPolicy
```

my Dockerfile is configured like so:

```DockerFile
FROM ubuntu:18.04
FROM rasa/rasa:2.5.1-full

USER root
RUN apt-get update && apt-get install -y python3.7 python3-pip
COPY . .
WORKDIR .

RUN chmod +x ./script
RUN pip3 install -r requirements.txt
USER 1001

EXPOSE ports-here
ENTRYPOINT ["sh", "script", "run"]
```

Let me know if you need anything else

**Comment by hsm207** [comment]
hey @jeff-ridgeway 

Could you try to reproduce the issue with mood bot? i.e. just run `rasa init` and then train the bot in SageMaker and then take the model.tar.gz file into an EC2 instance and set up mood bot using docker compose.

If the same issue appears, please share the model.tar.gz file.

**Comment by hsm207** [comment]
hey @jeff-ridgeway 

> Note: I ran the tar and echo commands and it said the version of rasa is 2.6.0. Does that mean that I'm running the wrong version? 🤔

Rasa 2.6.0 was released  a few days ago.

It sounds like the AWS SageMaker is set up to always pull the latest version of Rasa.

Anyway, now that you have a model that is clearly trained with Rasa 2.6.0, if you adjust your docker setup to also use Rasa 2.6.0, then it should load fine.

**Comment by jeff-ridgeway** [comment]
hey @hsm207 

So I think the version not being 2.6.0 was the culprit, however I'm not out of the woods yet. Now I get the following error when loading my original model after updating the rasa version to `2.6.0.` in the Dockerfile and building the docker image (that was the key) and modifying my docker-compose file to just used the built image. However, here is new the error:

```
rasa  | 2021-05-07 19:43:20 ERROR    rasa.core.agent  - Could not load model due to Please confirm that en_core_web_md is an available spaCy model. You need to download one upfront. For example:
rasa  | python -m spacy download en_core_web_md
rasa  | More informaton can be found on https://rasa.com/docs/rasa/components#spacynlp.
rasa  | /opt/venv/lib/python3.8/site-packages/rasa/shared/utils/io.py:97: UserWarning: The model at 'models/' could not be loaded. Error: <class 'rasa.nlu.model.InvalidModelError'>: Please confirm that en_core_web_md is an available spaCy model. You need to download one upfront. For example:
rasa  | python -m spacy download en_core_web_md
rasa  | More informaton can be found on https://rasa.com/docs/rasa/components#spacynlp
```

Any help would still be appreciated on this error, though I'm working on it. Thank you for the patience and helping me out thus far!

**Comment by hsm207** [comment]
hey @jeff-ridgeway ,

Based on the logs you shared, you're pipeline is relying on a Spacy model, namely `en_core_web_md`.

You need to manually download this model and link it.

Based on the setup you shared, you can modify your Dockerfile to include these lines:

```bash
RUN python -m spacy download en_core_web_md && \
python -m spacy download en_core_web_md
```

**Comment by jeff-ridgeway** [comment]
@hsm207 that was it! Thanks so much for all the help! I really really appreciate it!
